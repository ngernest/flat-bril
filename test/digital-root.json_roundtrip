{
  "functions": [
    {
      "args": [
        {
          "name": "input",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "ten",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "dest": "result",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "label": "begin"
        },
        {
          "args": [
            "input"
          ],
          "dest": "digit",
          "funcs": [
            "peel_last_digit"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "input",
            "ten"
          ],
          "dest": "input",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "result",
            "digit"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "label": "check_result"
        },
        {
          "args": [
            "result"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "args": [
            "result"
          ],
          "dest": "processed",
          "funcs": [
            "is_single_digit"
          ],
          "labels": [],
          "op": "call",
          "type": "bool"
        },
        {
          "args": [
            "processed"
          ],
          "funcs": [],
          "labels": [
            "check_done",
            "process_result"
          ],
          "op": "br"
        },
        {
          "label": "process_result"
        },
        {
          "args": [
            "result"
          ],
          "dest": "r0",
          "funcs": [
            "peel_last_digit"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "result",
            "ten"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "result",
            "r0"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "check_result"
          ],
          "op": "jmp"
        },
        {
          "label": "check_done"
        },
        {
          "args": [
            "input",
            "zero"
          ],
          "dest": "done",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "done"
          ],
          "funcs": [],
          "labels": [
            "done",
            "begin"
          ],
          "op": "br"
        },
        {
          "label": "done"
        },
        {
          "args": [
            "result"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "input",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "ten",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "input",
            "ten"
          ],
          "dest": "divided",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "divided",
            "ten"
          ],
          "dest": "mul_by_ten",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "mul_by_ten",
            "zero"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "result"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "is_single_digit",
      "type": "bool"
    },
    {
      "args": [
        {
          "name": "input",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "ten",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "args": [
            "input",
            "ten"
          ],
          "dest": "div_by_ten",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "div_by_ten",
            "ten"
          ],
          "dest": "mul_by_ten",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "input",
            "mul_by_ten"
          ],
          "dest": "last_digit",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "last_digit"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "peel_last_digit",
      "type": "int"
    }
  ]
}
