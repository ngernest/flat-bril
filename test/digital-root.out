@main
	op: const		dest: zero		type: int		value: 0	
	op: const		dest: ten		type: int		value: 10	
	op: const		dest: result		type: int		value: 0	
.begin
	op: call 		dest: digit		type: int		args: ["input"]		funcs: ["peel_last_digit"]
	op: div  		dest: input		type: int		args: ["input", "ten"]	
	op: add  		dest: result		type: int		args: ["result", "digit"]	
.check_result
	op: print		args: ["result"]	
	op: call 		dest: processed		type: bool		args: ["result"]		funcs: ["is_single_digit"]
	op: br   		args: ["processed"]		labels: ["check_done", "process_result"]
.process_result
	op: call 		dest: r0		type: int		args: ["result"]		funcs: ["peel_last_digit"]
	op: div  		dest: result		type: int		args: ["result", "ten"]	
	op: add  		dest: result		type: int		args: ["result", "r0"]	
	op: jmp  		labels: ["check_result"]
.check_done
	op: eq   		dest: done		type: bool		args: ["input", "zero"]	
	op: br   		args: ["done"]		labels: ["done", "begin"]
.done
	op: print		args: ["result"]	
@is_single_digit
	op: const		dest: ten		type: int		value: 10	
	op: const		dest: zero		type: int		value: 0	
	op: div  		dest: divided		type: int		args: ["input", "ten"]	
	op: mul  		dest: mul_by_ten		type: int		args: ["divided", "ten"]	
	op: eq   		dest: result		type: bool		args: ["mul_by_ten", "zero"]	
	op: ret  		args: ["result"]	
@peel_last_digit
	op: const		dest: ten		type: int		value: 10	
	op: div  		dest: div_by_ten		type: int		args: ["input", "ten"]	
	op: mul  		dest: mul_by_ten		type: int		args: ["div_by_ten", "ten"]	
	op: sub  		dest: last_digit		type: int		args: ["input", "mul_by_ten"]	
	op: ret  		args: ["last_digit"]	
